jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.cloud.tools.jib.filesystem.XdgDirectories;
+import com.google.cloud.tools.jib.json.JsonTemplate;
+import com.google.cloud.tools.jib.json.JsonTemplateMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/** Checks if Jib is up-to-date. */
+public class UpdateChecker {
+
+  /** JSON template for the configuration file used to enable/disable update checks. */
+  @VisibleForTesting
+  static class ConfigJsonTemplate implements JsonTemplate {
+    private boolean disableUpdateCheck;
+
+    @VisibleForTesting
+    void setDisableUpdateCheck(boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param skip if {@code true}, the update check itself will be skipped
+   * @param version the current version of Jib being used
+   * @param versionUrl the location to check for the latest version
+   * @param executorService the {@link ExecutorService}
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(
+      boolean skip, String version, String versionUrl, ExecutorService executorService) {
+    return new UpdateChecker(
```

Comment:
do you have do some work here to shutdown the executor?

Reply:
I should probably provide more context:

```
ExecutorService x = Executors.newSingleThreadExecutor();
x.submit(() -> System.out.println("goose"));
// x.shutdown();
```

In this example, without x.shutdown, running this will pause until the thread is cleaned up (or something like that)

More context here: https://stackoverflow.com/questions/20057497/program-does-not-terminate-immediately-when-all-executorservice-tasks-are-done

Reply:
Yeah, I think we can save the executor service and call `shutdown()` in `finishUpdateCheck()`.

Reply:
Oh, looks like we can shut it down immediately after submitting.

Reply:
Should the executor service be isolated to the update checker? I may have missed why we are passing it in?

Reply:
Yeah you can use shutdown if you know you're not submitting any more jobs.
But if the executor is coming from outside, then there's no way to ensure that it is exclusively being used for this purpose?